File: base/src/main/java/vjson/parser/DefaultParserCacheHolder.java
package vjson.parser;

public class DefaultParserCacheHolder implements ParserCacheHolder {
    private boolean started = false;

    private static final ThreadLocal<ArrayParser> threadLocalArrayParser = new ThreadLocal<>();
    private static final ThreadLocal<ObjectParser> threadLocalObjectParser = new ThreadLocal<>();
    private static final ThreadLocal<StringParser> threadLocalStringParser = new ThreadLocal<>();

    private static final ThreadLocal<ArrayParser> threadLocalArrayParserJavaObject = new ThreadLocal<>();
    private static final ThreadLocal<ObjectParser> threadLocalObjectParserJavaObject = new ThreadLocal<>();
    private static final ThreadLocal<StringParser> threadLocalStringParserJavaObject = new ThreadLocal<>();

    public boolean isStarted() {
        return started;
    }

    @Override
    public ArrayParser threadLocalArrayParser() {
        return threadLocalArrayParser.get();
    }

    @Override
    public void threadLocalArrayParser(ArrayParser parser) {
        started = true;
        threadLocalArrayParser.set(parser);
    }

    @Override
    public ObjectParser threadLocalObjectParser() {
        return threadLocalObjectParser.get();
    }

    @Override
    public void threadLocalObjectParser(ObjectParser parser) {
        started = true;
        threadLocalObjectParser.set(parser);
    }

    @Override
    public StringParser threadLocalStringParser() {
        return threadLocalStringParser.get();
    }

    @Override
    public void threadLocalStringParser(StringParser parser) {
        started = true;
        threadLocalStringParser.set(parser);
    }

    @Override
    public ArrayParser threadLocalArrayParserJavaObject() {
        return threadLocalArrayParserJavaObject.get();
    }

    @Override
    public void threadLocalArrayParserJavaObject(ArrayParser parser) {
        started = true;
        threadLocalArrayParserJavaObject.set(parser);
    }

    @Override
    public ObjectParser threadLocalObjectParserJavaObject() {
        return threadLocalObjectParserJavaObject.get();
    }

    @Override
    public void threadLocalObjectParserJavaObject(ObjectParser parser) {
        started = true;
        threadLocalObjectParserJavaObject.set(parser);
    }

    @Override
    public StringParser threadLocalStringParserJavaObject() {
        return threadLocalStringParserJavaObject.get();
    }

    @Override
    public void threadLocalStringParserJavaObject(StringParser parser) {
        started = true;
        threadLocalStringParserJavaObject.set(parser);
    }
}

File: base/src/main/kotlin/vjson/parser/DefaultParserCacheHolder.kt
/*
 * The MIT License
 *
 * Copyright 2021 wkgcass (https://github.com/wkgcass)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
// #ifndef KOTLIN_NATIVE {{
package vjson.parser

import vjson.util.StringDictionary

open class DefaultParserCacheHolder : ParserCacheHolder {
  var isStarted = false
    private set

  companion object {
    private val threadLocalArrayParser = ThreadLocal<ArrayParser>()
    private val threadLocalObjectParser = ThreadLocal<ObjectParser>()
    private val threadLocalStringParser = ThreadLocal<StringParser>()
    private val threadLocalArrayParserJavaObject = ThreadLocal<ArrayParser>()
    private val threadLocalObjectParserJavaObject = ThreadLocal<ObjectParser>()
    private val threadLocalStringParserJavaObject = ThreadLocal<StringParser>()
    private val threadLocalKeyDictionary = ThreadLocal<StringDictionary>()
  }

  override fun threadLocalArrayParser(): ArrayParser? {
    return threadLocalArrayParser.get()
  }

  override fun threadLocalArrayParser(parser: ArrayParser) {
    isStarted = true
    threadLocalArrayParser.set(parser)
  }

  override fun threadLocalObjectParser(): ObjectParser? {
    return threadLocalObjectParser.get()
  }

  override fun threadLocalObjectParser(parser: ObjectParser) {
    isStarted = true
    threadLocalObjectParser.set(parser)
  }

  override fun threadLocalStringParser(): StringParser? {
    return threadLocalStringParser.get()
  }

  override fun threadLocalStringParser(parser: StringParser) {
    isStarted = true
    threadLocalStringParser.set(parser)
  }

  override fun threadLocalArrayParserJavaObject(): ArrayParser? {
    return threadLocalArrayParserJavaObject.get()
  }

  override fun threadLocalArrayParserJavaObject(parser: ArrayParser) {
    isStarted = true
    threadLocalArrayParserJavaObject.set(parser)
  }

  override fun threadLocalObjectParserJavaObject(): ObjectParser? {
    return threadLocalObjectParserJavaObject.get()
  }

  override fun threadLocalObjectParserJavaObject(parser: ObjectParser) {
    isStarted = true
    threadLocalObjectParserJavaObject.set(parser)
  }

  override fun threadLocalStringParserJavaObject(): StringParser? {
    return threadLocalStringParserJavaObject.get()
  }

  override fun threadLocalStringParserJavaObject(parser: StringParser) {
    isStarted = true
    threadLocalStringParserJavaObject.set(parser)
  }

  override fun threadLocalKeyDictionary(): StringDictionary? {
    return threadLocalKeyDictionary.get()
  }

  override fun threadLocalKeyDictionary(dic: StringDictionary) {
    isStarted = true
    threadLocalKeyDictionary.set(dic)
  }
}
// }}
